ansible_collections/arista/avd/roles/eos_designs/python_modules/ip_addressing/__init__.py

```diff
@@ -0,0 +1,346 @@
+
+from collections import ChainMap
+from functools import cached_property
+import ipaddress
+import importlib
+from ansible_collections.arista.avd.plugins.plugin_utils.avdfacts import AvdFacts
+from ansible_collections.arista.avd.plugins.plugin_utils.utils import AristaAvdError, get
+
+
+DEFAULT_AVD_IP_ADDRESSING_PYTHON_MODULE = "ansible_collections.arista.avd.roles.eos_designs.python_modules.ip_addressing"
+DEFAULT_AVD_IP_ADDRESSING_PYTHON_CLASS_NAME = "AvdIpAddressing"
+
+
+class AvdIpAddressing(AvdFacts):
+
+    '''
+    Class used to render IP addresses either from custom Jinja2 templates or using default Python Logic
+
+    Since some templates might contain certain legacy variables (switch_*),
+    those are mapped from the switch.* model
+
+    This class is imported adhoc based on the variable `templates.ip_addressing.python_module` so it can
+    be overridden by a custom python class.
+    '''
+
+    def _ip(self, pool: str, prefixlen: int, subnet_offset: int, ip_offset: int) -> str:
+        network = ipaddress.ip_network(pool, strict=False)
+        subnets = network.subnets(new_prefix=prefixlen)
+        if isinstance(subnets, list):
+            # It seems that some versions of python return a list instead of the documented iterator
+            subnets = iter(subnets)
+        subnet_index = 0
+        try:
+            while subnet_index <= subnet_offset:
+                subnet = next(subnets)
+                subnet_index += 1
+        except StopIteration as e:
+            raise AristaAvdError(f"Unable to get {subnet_offset+1} /{prefixlen} subnets from pool {pool}") from e
```

Comment:
This could probably be replaced with something like (maybe need some verificatin at the boundaries):

```
subnets = list(network.subnets(new_prefix=prefixlen)
try:
  subnet = subnets[subnet_offset]
except OutOfRange:
   AristaAvdError ..
```

Reply:
I did as you suggest at first, but with a /64 in ipv6 it will try to produce a list with 64-bit entries, which takes quite a while :)
So instead I decided to use this iterator since we often need one of the first few hundred addresses. list() would be doing the same thing anyway, so there should be no cost to this. Only benefits when we pull the lower entries of the scope.